File: web/lib/kpi.ts
/**
 * The KPI contract.
 *
 * Every number YIELDMAP shows crosses the API as one of these objects, never as a bare float.
 * The point is Rule 3: a reader can always get from a rendered figure back to the exact SQL that
 * produced it, the sample it came from, and the method that defines it.
 */
import thresholds from "../../config/kpi_thresholds.json";

export type Confidence = "high" | "medium" | "low" | "insufficient";

export type KpiSource =
  | { kind: "sql"; ref: string }
  | { kind: "doc"; ref: string; sha256?: string };

export interface KpiDelta {
  value: number;
  /** e.g. "12m", "yoy", "30d" */
  window: string;
  direction: "up" | "down" | "flat";
}

export interface Kpi {
  id: string;
  label: string;
  /** null means the value could not be computed — render the insufficient-data state. */
  value: number | null;
  unit: string;
  format: "currency_aed" | "percent" | "count" | "ratio" | "index" | "score" | "years";
  delta?: KpiDelta;
  /** Sample size behind the value. Always present, always shown. */
  n: number;
  asof: string;
  confidence: Confidence;
  /** Deep-links to /methodology#<method_id>. */
  method_id: string;
  sql: string;
  sql_hash: string;
  sources: KpiSource[];
  /** Set when provenance is SYNTHETIC, so the UI can banner it. */
  synthetic?: boolean;
}

/**
 * Confidence from sample size and dispersion.
 *
 * `cv` is the coefficient of variation of the underlying sample (std / |mean|). A small sample or
 * a wildly dispersed one is not reported as a confident number — this is what stops the 3D city
 * from extruding a whole community on the strength of three transactions.
 */
export function confidenceFor(n: number, cv: number | null): Confidence {
  if (!Number.isFinite(n) || n < thresholds.min_n) return "insufficient";
  // Dispersion unknown (e.g. a count, which has none) — judge on sample size alone.
  const dispersion = cv === null || !Number.isFinite(cv) ? 0 : Math.abs(cv);
  if (n >= thresholds.high.min_n && dispersion <= thresholds.high.max_cv) return "high";
  if (n >= thresholds.medium.min_n && dispersion <= thresholds.medium.max_cv) return "medium";
  return "low";
}

/** True when the tile must render "insufficient data" instead of a number. */
export function isSuppressed(kpi: Pick<Kpi, "value" | "confidence">): boolean {
  return kpi.value === null || kpi.confidence === "insufficient";
}

const AED = new Intl.NumberFormat("en-AE", {
  style: "currency",
  currency: "AED",
  maximumFractionDigits: 0,
});

/** Display string for a KPI. Never invents a value: suppressed KPIs render a dash. */
export function formatKpi(kpi: Pick<Kpi, "value" | "format" | "unit" | "confidence">): string {
  if (isSuppressed(kpi)) return "—";
  const v = kpi.value as number;
  switch (kpi.format) {
    case "currency_aed":
      return AED.format(v);
    case "percent":
      return `${v.toFixed(2)}%`;
    case "count":
      return new Intl.NumberFormat("en-AE").format(Math.round(v));
    case "ratio":
      return v.toFixed(2);
    case "index":
      return v.toFixed(1);
    case "score":
      return Math.round(v).toString();
    case "years":
      return `${v.toFixed(1)} yr`;
  }
}

File: api/kpi.py
"""The KPI contract, Python side.

Mirrors ``web/lib/kpi.ts``. Both read ``config/kpi_thresholds.json`` so the API and the UI can
never disagree about whether a cell has enough data to be shown.

Rule 3 lives here: a ``Kpi`` cannot be constructed without the SQL that produced it, so there is no
code path that emits a number the user cannot trace back to a query.
"""

from __future__ import annotations

import hashlib
import json
import math
from datetime import date
from pathlib import Path
from typing import Literal

from pydantic import BaseModel, Field, field_validator

_THRESHOLDS_PATH = Path(__file__).resolve().parent.parent / "config" / "kpi_thresholds.json"
THRESHOLDS: dict = json.loads(_THRESHOLDS_PATH.read_text())

Confidence = Literal["high", "medium", "low", "insufficient"]
KpiFormat = Literal["currency_aed", "percent", "count", "ratio", "index", "score", "years"]
Provenance = Literal["REAL", "SYNTHETIC"]


def confidence_for(n: int, cv: float | None) -> Confidence:
    """Confidence from sample size and dispersion.

    ``cv`` is the coefficient of variation of the underlying sample (std / |mean|), or ``None`` for
    a quantity that has no dispersion, such as a count. A small or wildly dispersed sample is never
    reported confidently — this is what stops the 3D city extruding a community on three sales.
    """
    if n is None or not isinstance(n, int) or n < THRESHOLDS["min_n"]:
        return "insufficient"
    dispersion = 0.0 if cv is None or not math.isfinite(cv) else abs(cv)
    if n >= THRESHOLDS["high"]["min_n"] and dispersion <= THRESHOLDS["high"]["max_cv"]:
        return "high"
    if n >= THRESHOLDS["medium"]["min_n"] and dispersion <= THRESHOLDS["medium"]["max_cv"]:
        return "medium"
    return "low"


def sql_hash(sql: str) -> str:
    """Stable short hash of a query, used to link a rendered number to its source."""
    return hashlib.sha256(" ".join(sql.split()).encode()).hexdigest()[:16]


class KpiDelta(BaseModel):
    value: float
    window: str
    direction: Literal["up", "down", "flat"]


class KpiSource(BaseModel):
    kind: Literal["sql", "doc"]
    ref: str
    sha256: str | None = None


class Kpi(BaseModel):
    """A single number, plus everything needed to check it."""

    id: str
    label: str
    value: float | None
    unit: str
    format: KpiFormat
    n: int = Field(ge=0)
    asof: date
    method_id: str
    sql: str
    sql_hash: str = ""
    confidence: Confidence = "insufficient"
    delta: KpiDelta | None = None
    sources: list[KpiSource] = Field(default_factory=list)
    provenance: Provenance = "REAL"

    @field_validator("sql")
    @classmethod
    def _sql_must_be_real(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("a Kpi must carry the SQL that produced it (project Rule 3)")
        return v

    @property
    def synthetic(self) -> bool:
        return self.provenance == "SYNTHETIC"

    @property
    def suppressed(self) -> bool:
        """True when the tile must show 'insufficient data' rather than a number."""
        return self.value is None or self.confidence == "insufficient"


def make_kpi(
    *,
    id: str,
    label: str,
    value: float | None,
    unit: str,
    format: KpiFormat,
    n: int,
    asof: date,
    method_id: str,
    sql: str,
    cv: float | None = None,
    delta: KpiDelta | None = None,
    sources: list[KpiSource] | None = None,
    provenance: Provenance = "REAL",
) -> Kpi:
    """Build a Kpi, deriving confidence and the SQL hash.

    Values that fail the confidence floor are blanked here rather than at render time, so a
    too-thin number never leaves the API in the first place.
    """
    conf = confidence_for(n, cv)
    if value is not None and not math.isfinite(value):
        value = None
    srcs = list(sources) if sources else [KpiSource(kind="sql", ref=method_id)]
    return Kpi(
        id=id,
        label=label,
        value=None if conf == "insufficient" else value,
        unit=unit,
        format=format,
        n=n,
        asof=asof,
        method_id=method_id,
        sql=sql,
        sql_hash=sql_hash(sql),
        confidence=conf,
        delta=delta,
        sources=srcs,
        provenance=provenance,
    )
